design/data-flow.md

```diff
@@ -0,0 +1,68 @@
+                        - Check for duplicate 
+                            - inputs
+                            - nullifiers (within a single transaction)
+                            - // Transactions containing empty `vin` must have either non-empty `vJoinSplit` or non-empty `vShieldedSpend`.
+                            - // Transactions containing empty `vout` must have either non-empty `vJoinSplit` or non-empty `vShieldedOutput`.
+                            - Moar: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L1091
+        - Sum up "LegacySigOps" for each transaction and check that it's less than some maximum 
+    - Acquires a lock, then calls `MarkBlockAsReceived` (networking?)
+    - Calls `AcceptBlock`, defined at: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4180
+        - Calls `AcceptBlockHeader`, defined at: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4130
+            - Also calls `CheckBlockHeader`
+            - Calls `ContextualCheckBlockHeader`, defined at: https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L3900
+                - Does checks given a pointer to the previous block
+                - Check Equihash solution is valid
+                    - In our code we compute the equihash solution on the block alone, we will need to also do a step to check that its block height is the appopriate N+1 re: the previous block
+                - Check proof of work
+                - Check timestamp against prev
+                - Check future timestamp soft fork rule introduced in v2.1.1-1.
+                    - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4027
+                - Check timestamp
+                    - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4040
+                - Reject block.nVersion < 4 blocks
+                - Don't accept any forks from the main chain prior to last checkpoint
+                    - We will probably get this 'more naturally' and don't need to explicitly check it like this
+        - Calls `CheckBlock` also, w/ ProofVerifier::Disabled == true
+            - https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4211
+        - Calls `ContextualCheckBlock`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L4065
+            - For each transaction:
+                - Calls `ContextualCheckTransaction`, defined at https://github.com/zcash/zcash/blob/ab2b7c0969391d8a57d90d008665da02f3f618e7/src/main.cpp#L760
+                    - For Zebra, we should only be doing Transaction v4 (Sapling+) checks, because checkpointing
```

Comment:
Do we need to keep context from v1-v3 transactions, so we can verify v4 transactions?

For example, UTXOs and revealed nullifiers.

Reply:
So, we might have to compute the treestate including nullifier set that is correct at the final checkpoint?

![image](https://user-images.githubusercontent.com/552961/89376641-39cba880-d6be-11ea-8614-d2e5e969b699.png)

Reply:
To answer your question, "I'm not sure"

![image](https://user-images.githubusercontent.com/552961/89378199-51586080-d6c1-11ea-9ced-58011261433c.png)